x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -704,18 +701,34 @@ public Object visitInputParam(EsqlBaseParser.InputParamContext ctx) {
         }
     }
 
-    private TypedParamValue param(TerminalNode node) {
+    Param paramByToken(TerminalNode node) {
         if (node == null) {
             return null;
         }
-
         Token token = node.getSymbol();
-
-        if (params.containsKey(token) == false) {
+        if (params.containsTokenLocation(token) == false) {
             throw new ParsingException(source(node), "Unexpected parameter");
         }
-
-        return params.get(token);
+        return params.getParamByTokenLocation(token);
     }
 
+    Param paramByNameOrPosition(TerminalNode node) {
+        if (node == null) {
+            return null;
+        }
+        Token token = node.getSymbol();
+        String nameOrPosition = token.getText().substring(1);
+        if (nameOrPosition.matches("[0-9]+")) {
```

Comment:
Use regex only when necessary due to their runtime cost.
Likely there's such a string utility in ES - otherwise just check each char through `Character.isLetterOrDigit` .